[SYCL][AOT][NVPTX][AMD][New offloading model] Add AOT support for NVIDIA and AMD backend in the clang-linker-wrapper for SYCL offloading

asudarsa opened the pull request:
This PR adds third-party (nvptx, amd) AOT support for SYCL offloading in the clang-linker-wrapper tool.
Following list of changes are found;

1. Code inside clang-linker-wrapper has been refactored to get it ready for non-SPIRV backends for SYCL offloading. Main changes are inside linkAndWrapDeviceFiles. Explanations are added as PR comments.
2. A new clang-linker-wrapper option to pass additional device library files for NVPTX backend has been added
3. An additional check to emit error when device library files are unexpectedly empty has been added
4. New tests have been added to test the compilation flow for the new changes and also existing tests have been modified to react to the additional check correctly.

E2E tests and more extensive driver tests will be added in the next stage.

Thanks

asudarsa commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -11084,9 +11084,7 @@ void LinkerWrapper::ConstructJob(Compilation &C, const JobAction &JA,
-      // TODO: Third party AOT support needs to be added in new offloading
-      // model.
-      if (TC->getTriple().isSPIROrSPIRV()) {
+      if (TC->getTriple().isSPIROrSPIRV() || TC->getTriple().isNVPTX()) {
```
For AMD, we do not pass any SYCL device libraries.

bader replied:
This comment seems useful for code readers not just PR reviewers.

asudarsa replied:
- Ah..you are right. Let me take a look at all these comments and move them into the code if it seems like a better fit.
- Thanks so much for looking closer.

Sincerely

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -768,21 +768,13 @@ getTripleBasedSPIRVTransOpts(const ArgList &Args,
 
 // Run LLVM to SPIR-V translation.
-static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef InputTable,
+static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef File,
```
So far, this function and runAOTCompile function have been reading in from the SYCL table file and writing back to it. They have been modified to now work on individual inputs. The reading/writing of SYCL table file have been moved to a higher level of call hierarchy.

bader replied:
It makes sense to extend the code comments before the function:

"Converts `File` file from LLVM bitcode to SPIR-V format using `llvm-spirv` tool. `Args` argument provides additional arguments for `llvm-spirv` tool."

Similar descriptions will be useful for all modified methods below.

asudarsa replied:
Will add the comments. Thanks.

asudarsa replied:
I have made a full parse of the file and added detailed comments for all SYCL specific functions.

Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -833,9 +814,8 @@ static void addBackendOptions(const ArgList &Args,
-static Error runAOTCompileIntelCPU(SmallVector<std::string, 16> InputFiles,
-                                   SmallVector<std::string, 16> OutputFiles,
-                                   const ArgList &Args) {
+static Expected<StringRef> runAOTCompileIntelCPU(StringRef InputFile,
```
Please see comments to runLLVMToSPIRVTranslation for explanation

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -847,31 +827,22 @@ static Error runAOTCompileIntelCPU(SmallVector<std::string, 16> InputFiles,
-static Error runAOTCompileIntelGPU(SmallVector<std::string, 16> &InputFiles,
-                                   SmallVector<std::string, 16> &OutputFiles,
-                                   const ArgList &Args) {
+static Expected<StringRef> runAOTCompileIntelGPU(StringRef InputFile,
```
Please see comments to runLLVMToSPIRVTranslation for explanation

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -889,54 +860,37 @@ static Error runAOTCompileIntelGPU(SmallVector<std::string, 16> &InputFiles,
 
 /// Run AOT compilation
-static Expected<StringRef> runAOTCompile(StringRef InputTable,
+static Expected<StringRef> runAOTCompile(StringRef InputFile,
```
Please see comments to runLLVMToSPIRVTranslation for explanation

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1237,6 +1191,35 @@ static Expected<StringRef> linkDevice(ArrayRef<StringRef> InputFiles,
+    }
+  }
+
+  if (ExtractedDeviceLibFiles.empty()) {
```
Extra check added to make sure that device library files are available when expected.

bader replied:
```suggestion
  // Make sure that device library files are available.
  if (ExtractedDeviceLibFiles.empty()) {
```

asudarsa replied:
Will add this to the upcoming commit. Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -2055,83 +2030,110 @@ Expected<SmallVector<StringRef>> linkAndWrapDeviceFiles(
-
     if (HasSYCLOffloadKind) {
-      // Link the remaining device files using the device linker for SYCL
+      SmallVector<StringRef> InputFiles;
```
Design so far worked well for SYCL + SPIRV target. For non-spirv targets, we need to call clang instead of spirv-translator+AOTCompile on each of the split file output of sycl-post-link. Hence sycl-post-link and parsing of it table output has been moved here.
Initial validation has been performed with E2E tests on local third party hardware.

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1393,10 +1376,7 @@ Expected<StringRef> linkDevice(ArrayRef<StringRef> InputFiles,
-      if (!SYCLPostLinkFile)
-        return SYCLPostLinkFile.takeError();
-      auto SPVFile = sycl::runLLVMToSPIRVTranslation(*SYCLPostLinkFile, Args);
+      auto SPVFile = sycl::runLLVMToSPIRVTranslation(InputFiles[0], Args);
```
sycl-post-link has been moved up as it is required for SYCL offloading for non-spirv targets as well.

bader commented:
> 1. Explanations are added as PR comments.

Why not **_code_** comments?

bader commented on clang/test/Driver/Inputs/libsycl-complex-amd.new.o:
Please, commit source files. Binary files are unmaintainable.

asudarsa replied:
These tests are placeholder tests and will disappear as soon as we add end-to-end tests (within the next month).
Please let me know if that's agreeable.

Thanks

bader replied:
I don't think "placeholder tests" is a good idea in general. I suggest we drop "placeholder tests" and commit only final tests.

Regarding end-to-end tests: it's better to have standalone tests for corresponding component compiler driver or clang-linker-wrapper. Running and debugging compiler components with end-to-end tests is much more difficult as it requires additional dependencies like SYCL runtime library or specialized hardware.

asudarsa replied:
I looked around some more into our codebase. Seems like I might be able to generate these .o files on the fly. Let me give it a shot.
Thanks

asudarsa replied:
Removed all .o files.

Thanks for the feedback.

asudarsa commented:
> > 1. Explanations are added as PR comments.
> 
> Why not **_code_** comments?

These comments are meant for reviewers who have been involved in the incremental development of this feature support. It explains what existed in the previous version and how things evolved in this version as I added AOT support for third party hardware. These comments do not make much sense for users/developers looking at the final version of the code. Hope that is agreeable.

Thanks

bader reviewed (commented):
>These comments are meant for reviewers who have been involved in the incremental development of this feature support. It explains what existed in the previous version and how things evolved in this version as I added AOT support for third party hardware. These comments do not make much sense for users/developers looking at the final version of the code. Hope that is agreeable.

Some of these comments for reviewers is better to have in code.

mdtoguchi commented:
overall - looks OK to me.  Given there are still some changes needed to improve the testing area, I'll have @hchilama take another look when the bits are ready as she is covering for me while I'm out.

asudarsa commented:
> overall - looks OK to me. Given there are still some changes needed to improve the testing area, I'll have @hchilama take another look when the bits are ready as she is covering for me while I'm out.

I have made changes to driver tests that I have added over time (mainly for new offloading model testing) to remove all .o files and am now generating them on the fly. @hchilama, can you please take a look when convenient?

Thanks

hchilama reviewed (approved):
LGTM - Driver

frasercrmck commented on clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:
```diff
@@ -150,6 +150,9 @@ def sycl_device_lib_EQ : CommaJoined<["--", "-"], "sycl-device-libraries=">,
 def sycl_device_library_location_EQ : Joined<["--", "-"],
   "sycl-device-library-location=">, Flags<[WrapperOnlyOption]>,
   HelpText<"Location of SYCL device library files">;
+def sycl_nvptx_device_lib_EQ : CommaJoined<["--", "-"], "sycl-nvptx-device-libraries=">,
```
Note sure we should have `nvptx`-specific options if we can avoid it. It makes it harder to update if anything changes in the future with supported SYCL targets, or with how we compile AMD.

Would it be at all possible to link this option up with `-Xsycl-target-backend`? Something like `-Xsycl-target-backend -sycl-device-libraries=...`? Then if there were multiple targets you could specify `-Xsycl-target-backend=nvptx64-nvidia-cuda -sycl-device-libraries=... -Xsycl-target-backend=amdgcn-amd-amdhsa -sycl-device-libraries=...`?

asudarsa replied:
Hi @frasercrmck 

Thanks so much for taking a look. One of the caveats in this PR is that this is not expected to be the final version to be upstreamed. This is expected to be an intermediate solution that is expected to functionally work. One of the key steps we intend t perform before upstreaming will be to refactor how we have introduced non-generic clang-linker-wrapper options. About this particular option, I feel that the we might be able to replace this option completety.

Hope it's agreeable from this PR's perspective.

Sincerely

frasercrmck replied:
Ah okay, I wasn't aware of that. That's fine, then. I just worry about the impact of changing flags once there's code out there that might be relying on them.

frasercrmck commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1194,6 +1196,13 @@ linkDeviceLibFiles(SmallVectorImpl<StringRef> &InputFiles,
   return *OutFileOrErr;
 }
 
+/// This function is used to link all SYCL device input files into a single
```
Could we specify all of this new documentation with doxygen, like the rest of LLVM/clang? Then we should be able to get properly rendered documentation.

asudarsa replied:
I am adding detailed code comments. Again, this is another step that we intend to perform as part of pre-upstreaming effort. Thanks

frasercrmck replied:
Okay, no problem.

frasercrmck commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -889,54 +890,41 @@ static Error runAOTCompileIntelGPU(SmallVector<std::string, 16> &InputFiles,
-    return *Output;
+    if (Triple.getSubArch() == llvm::Triple::SPIRSubArch_gen) {
+      auto GPUOutputFile = runAOTCompileIntelGPU(InputFile, Args);
+      if (!GPUOutputFile)
```
Do we really need to perform this step - can't we just `return runAOTCompileIntelGPU(InputFile, Args)` directly? We're just forwarding on a `Expected<StringRef>` from one function to another.

asudarsa replied:
Good point. I will update. Thanks

bader commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -495,10 +495,15 @@ static Error getSYCLDeviceLibs(SmallVector<std::string, 16> &DeviceLibFiles,
+/// If input is not a SPIR-V file, then the original file is returned.
+/// TODO: Add a check to identify SPIR-V files and exit early if the input is
+/// not a SPIR-V file.
+/// 'Filename' is the input file that could be a SPIR-V file.
+/// 'Args' encompasses all arguments required for linking and wrapping device
+/// code and will be parsed to generate options required to be passed into the
+/// llvm-spirv tool.
```
If we want to apply proper Doxygen annotations, we should add `\p` before the parameter name (or `\param` in extended form). See https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments for code examples:
```suggestion
/// \param Filename is the input file that could be a SPIR-V file.
/// \param Args encompasses all arguments required for linking and wrapping device
/// code and will be parsed to generate options required to be passed into the
/// llvm-spirv tool.
```

asudarsa replied:
I am hoping to wrap up this PR and do a 'global' comments update (for all my clang-linker-wrapper changes) as a follow-up step. Hope that's agreeable.

Thanks

bader replied:
It will be cleaner to fix annotations for the comments added in this PR to avoid dependency on future changes (i.e. make this change self-contained). I don't want this to block the merge, so I'll approve and leave the decision on how to proceed with this pull request to @asudarsa and the other reviewers.

bader commented on clang/test/Driver/Inputs/test-sycl.o:
👍 
It's a pity that we have almost 1Mb of unnecessary binary files in the git history.

asudarsa replied:
Ah..you are right......sorry about that.....:-(

bader commented on clang/test/Driver/sycl-post-link-options-win.cpp:
```diff
@@ -4,13 +4,26 @@
+//
 // RUN: clang-linker-wrapper --dry-run --host-triple=x86_64-pc-windows-msvc \
-// RUN:   -sycl-device-library-location=%S/Inputs -sycl-device-libraries=libsycl-crt.new.obj \
+// RUN:   -sycl-device-library-location= -sycl-device-libraries=%t.devicelib.o \
```
```suggestion
// RUN:   -sycl-device-libraries=%t.devicelib.o \
```
Do we need setting `-sycl-device-library-location` option?

asudarsa replied:
I had a check in my linker-wrapper code to look for empty library location and throw an error if it is not set. I think I do  not need to do that. Removed it.

Thanks

asudarsa commented on clang/test/Driver/linker-wrapper-sycl-win.cpp:
```diff
@@ -1,40 +1,139 @@
 /// Check for list of commands for standalone clang-linker-wrapper run for sycl
-// RUN: clang-linker-wrapper -sycl-device-library-location=%S/Inputs -sycl-device-libraries=libsycl-crt.new.obj,libsycl-complex.new.obj -sycl-post-link-options="SYCL_POST_LINK_OPTIONS" -llvm-spirv-options="LLVM_SPIRV_OPTIONS" "--host-triple=x86_64-pc-windows-msvc" "--linker-path=/usr/bin/ld" "--" HOST_LINKER_FLAGS "-dynamic-linker" HOST_DYN_LIB "-o" "a.out" HOST_LIB_PATH HOST_STAT_LIB %S/Inputs/t ...
+// -------
+// Generate .o file as linker wrapper input.
```
Yes. We require some .o file with a valid device image. Driver is more or less 'static'. But the linker-wrapper is dynamic and will skip all steps if no valid device image is found.

Thanks

asudarsa commented on clang/test/Driver/linker-wrapper-sycl-win.cpp:
```diff
@@ -1,40 +1,139 @@
+// RUN: %clang -cc1 %s -triple spir64-unknown-unknown -aux-triple x86_64-pc-windows-msvc -fsycl-is-device -emit-llvm-bc -o %t.bc
+// RUN: clang-offload-packager -o %t.out --image=file=%t.bc,kind=sycl,triple=spir64-unknown-unknown
+// RUN: %clang -cc1 %s -triple x86_64-pc-windows-msvc -emit-obj -o %t.o \
+// RUN:   -fembed-offload-object=%t.out
```
I think you are correct. I tried it locally and it does work as expected. We do have to pass --offload-new-driver as well to make sure packager is called and not bundler.

bader replied:
Right. I suggest we call the driver. It will be just one simple command instead of multiple commands.

asudarsa commented on clang/test/Driver/linker-wrapper-sycl-win.cpp:
```diff
@@ -1,40 +1,139 @@
+// -------
+// Generate .o file as linker wrapper input.
+//
+// RUN: %clang -cc1 %s -triple spir64-unknown-unknown -aux-triple x86_64-pc-windows-msvc -fsycl-is-device -emit-llvm-bc -o %t.bc
```
I am replacing all such commands with the top-level %clang command. So, we can ignore this comment.

asudarsa commented on clang/test/Driver/linker-wrapper-sycl-win.cpp:
```diff
@@ -1,40 +1,139 @@
+//
+// RUN: echo '' > %t.devicelib.cpp
+// RUN: %clang -cc1 %t.devicelib.cpp -triple spir64-unknown-unknown -aux-triple x86_64-pc-windows-msvc -fsycl-is-device -emit-llvm-bc -o %t.devicelib.bc
+// RUN: clang-offload-packager -o %t.devicelib.out --image=file=%t.devicelib.bc,kind=sycl,triple=spir64-unknown-unknown
```
I am no longer requiring a .bc fil e as I have moved to top-level command. I will use 'touch' command for the .cpp file.

Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -2055,83 +2079,110 @@ Expected<SmallVector<StringRef>> linkAndWrapDeviceFiles(
+      if (!OutputOrErr)
+        return OutputOrErr.takeError();
+      // Store the offloading image for each linked output file.
+      for (OffloadKind Kind : ActiveOffloadKinds) {
+        llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> FileOrErr =
```
Use case is this: If we have t1.bc (cuda offloading, x86_64 target) and t2.bc (openmp offloading, x86_64 target), then t1.bc and t2.bc will be linked together  and then following images will be created: t.cuda.bc and t.openmp.bc

If we had SYCL offloading for t1.bc, then we will link the two files using our approach to create t.sycl.bc using our approach and link the two files using community approach to create t.openmp.bc

I am not seeing any red flags here. But, I do not have realistic use cases for these. I can add them as part of my E2E testing.

sarnex replied:
Ah ok, I didn't think we would want to link them together.

sarnex reviewed (approved):
LGTM as incremental change as long as feedback is tracked and addressed in follow-up PR

asudarsa commented:
Hi @bader 

Thanks so very much for the detailed review. I have learnt a bunch about writing better tests here. 
I have addressed all the comments except the one related to 'doxygen' documentation which will be addressed in an upcoming patch.

Sincerely

maksimsab reviewed (approved):
clang-linker-wrapper part LGTM.

Naghasan reviewed (commented):
@asudarsa I'm in agreement with Fraser about his comments but obviously everyone is. Just had a small minor comment. Change looks good, thanks.
